### DISCUSSION TITLE
호이스팅이란 무엇이며, 실행 컨텍스트와 어떤 관련이 있나요?

### DISCUSSION THREAD

**Discussion by publdaze** [publdaze/js-study#72] [실행컨텍스트]
- 호이스팅이란 무엇이며, 실행 컨텍스트와 어떤 관련이 있나요?
  - 변수 호이스팅과 함수 호이스팅의 차이점은 무엇인가요?
  - 코드를 보고 결과를 설명해주세요.
    ```js
    console.log(abcd)
    abcd()
    var abcd = () => {
        console.log("abcd");
    }
    ```

**Comment by publdaze** [comment]
> - 호이스팅이란 무엇이며, 실행 컨텍스트와 어떤 관련이 있나요?
- 호이스팅은 스코프의 최상단으로 변수나 함수 선언문이 끌어올려지는 것처럼 동작하는 현상을 말함.
- 코드 평가 단계에서 실행 컨텍스트가 생성이 되고, 렉시컬 환경의 환경 레코드에 식별자가 저장이 될 것
- 이때 var 키워드로 선언된 변수의 경우는 호이스팅이 일어나며, 선언과 초기화 단계가 동시에 일어나기 때문에 undefined로 값이 할당되고,
- let, const 키워드로 선언된 변수의 경우는 마찬가지로 호이스팅이 일어나지만, 선언 단계가 먼저 수행되고, 런타임의 초기화 시점에 초기화 단계가 일어나므로 그 사이에 존재하는 TDZ에서는 초기화되지 않은 상태로 남아있게 되며, 이때 해당 식별자를 참조할 시 reference에러가 발생
- 따라서 실행 컨텍스트가 생성이 될 때 호이스팅이 일어난 식별자들을 실행 전에 저장해둘 수 있다는 측면에서 호이스팅과 실행 컨텍스트가 서로 관련이 있다고 할 수 있음
> - 변수 호이스팅과 함수 호이스팅의 차이점은 무엇인가요?
- 변수 호이스팅은 변수 선언만 호이스팅되어 선언 전 변수 참조시 var의 경우 undefined, let/const의 경우 reference에러 발생
- 함수 호이스팅은 함수 선언문 전체가 호이스팅되어 함수 선언 전에 호출이 가능
> - 코드를 보고 결과를 설명해주세요.
>   ```js
>   console.log(abcd)
>   abcd()
>   var abcd = () => {
>       console.log("abcd");
>   }
>   ```
- abcd 화살표 함수는 var 변수로 선언되어 있음
- var의 경우는 코드 평가 단계에 호이스팅이 되고, 선언과 동시에 undefined로 초기화 되므로 첫 출력은 undefined가 될 것
- 이후 abcd 호출이 일어나는데, abcd가 함수 선언문이었으면 정의까지 호이스팅되어 함수 호출이 가능했겠지만, 함수 정의식으로 함수가 선언되어 있기 때문에 에러(abcd가 함수가 아니라는 type에러)가 발생함

**Comment by pipisebastian** [comment]
> 호이스팅이란 무엇이며, 실행 컨텍스트와 어떤 관련이 있나요?

1. **호이스팅**이 끌어올리다. 라는 뜻인데, 이게 변수나 함수의 선언이, 마치  코드 최상단에서 이뤄지는것처럼 보이는 현상입니다. 그래서 선언 line 이전에 변수를 참조하더라도, 에러가 뜨지 않는데요. (var, 함수선언문만 해당)
2. 그러나 let, const 의 경우에는, 마찬가지로 호이스팅이 일어나나, 보기에는 호이스팅이 일어나지 않는 것처럼 보입니다. 그 이유는, let,const는 선언과 초기화단계가 분리되어 일어나는데, 초기화 단계는 코드상에서 할당되는 시점에 일어납니다. 그렇기에 var처럼 에러가 안뜨지는 않고, reference error가 뜨게 됩니다.
3. 이게 실제로 코드 최상단에 선언이 끌어올려지는건 아니고, 함수가 최초로 실행될 때, 실행컨텍스트의 lexical 환경 속, `enviorment Record`에 식별자 정보를 저장했기 때문에, 참조가 가능하게 되는 겁니다.

> 변수 호이스팅과 함수 호이스팅의 차이점은 무엇인가요?

- 변수는 선언문 이전에도 참조가 가능하긴 하지만, undefined를 출력합니다.
- 그러나, 함수 선언문의 경우, 소스코드 평가 단계에서, 함수 객체를 즉시 할당합니다. ( == 함수 전체를 호이스팅한다. 즉, 함수의 선언/정의가 모두 호이스팅) 그렇기에, 선언문 이전에도 함수를 호출 가능합니다. 

> 코드를 보고 결과를 설명해주세요

- undefined, TypeError
- abcd  함수는 함수 선언문이 아니기 때문에, 변수 호이스팅이 발생합니다.
- 그런데, var키워드로 선언되었기 때문에, 호이스팅이 일어나, undefined가 출력합니다.
- 그러나, abcd를 호출한 상태에서는, 아직 abcd는 undefined이기 때문에, 이를 호출하기에 TypeError가 발생합니다.

**Reply by pipisebastian** [reply]
함수 호이스팅 공부 필요!!! 🚨
